Update troubleshooting doc (#891)
superfly/docs: getting-started/troubleshooting.html.md

---
title: Troubleshooting your deployment
layout: docs
sitemap: false
nav: firecracker
---

> As a developer, you probably know that there are many reasons why an app deployment can fail. This section gives you some ideas of where to start, but it's not a comprehensive guide. If you're still stuck after reading, then visit our [community forum](https://community.fly.io/) for more help.

Author: we are all developers here

## Try this first

If the error you get isn’t obvious or specific, and the answer doesn’t appear right away, then try these basic steps first, to either fix the problem or to arm yourself with knowledge.

### Update flyctl

By default, flyctl (the Fly CLI), [updates automatically](https://community.fly.io/t/flyctl-versions-autoupdating-and-the-cli-apocalypse/13794).

But if you disabled automatic updates, then you should update flyctl:

```cmd
fly version update
```

> We frequently add new features to flyctl, so you should keep it up-to-date to avoid breaking things. You can also turn automatic updates back on with:

Author: `<embarrassed emoji face> `

## Get more information about failures

Logs have knowledge.

### Check the logs

Check the logs of the app when it's running or when you run `fly deploy`. Run your command in one terminal window and open a second window to view the logs.

To get the most recent log entries:

```cmd
fly logs
```

Look for error messages that indicate why the app or deploy is failing, and the logs that occurred just before the app crashed or the deploy failed.

If you can see messages about the app just exiting, then there's likely a specific app issue, and you'll need to address that before you can redeploy.

### Activate debug logging

Activate debug level logs on a command, like `fly deploy`:

```cmd
LOG_LEVEL=debug fly <command>
```

LOG_LEVEL=debug prints all the logs into the console as the command runs.

### Inspect with SSH

You can use `fly ssh console` to connect to a running instance of your application. Use `fly ssh console -s` to select a specific Machine.

## WARNING The app is listening on the incorrect address (Host and port checking)

> To be reachable by Fly Proxy, an app needs to listen on 0.0.0.0 and bind to the `internal_port` defined in the `fly.toml` configuration file.

Author: I think so

If your app is not listening on the expected address and the configured port, you’ll get the following warning message when you deploy your app:

```
WARNING The app is listening on the incorrect address and will not be reachable by fly-proxy.
```

The message supplies:

- The host address your app should be listening on, which is `0.0.0.0:<internal_port value>`.
- A list of processes inside the Machine with TCP sockets in LISTEN state. This includes `/app`, which might show something like `[ :: ]:8080`; the host address your app is trying to listen on. (You can ignore hallpass on port 22, which is used for SSH on Machines.)

### Fix the "app is listening on the incorrect address" error

The default internal port when you launch a new Fly App is `8080`.

You can either:
- Configure your app to listen on host `0.0.0.0:8080`, or
- Configure your app to listen on host `0.0.0.0:<framework default or other port>` and change the `internal_port` value in the `fly.toml` configuration file to match.

### Why does my app listen on localhost with a different port number?

> A lot of frameworks will listen on the localhost/127.0.0.1 by default so that the developer can connect to the app. Different frameworks also define different default ports, like 3000, 8000, or 8080, for example. It can be easy to make a mistake and configure your app in a way that makes it impossible for the Fly Proxy to route requests to it. And it can be difficult to debug, especially if your framework doesn't print the listening address to logs and your image doesn't have `netstat` or `ss` tools.

Reviewer: At this spot you could link to, or copy over, the info in /docs/getting-started/app-services (smallest scope is link to it)

> A lot of frameworks will listen on the localhost/127.0.0.1 by default so that the developer can connect to the app. Different frameworks also define different default ports, like 3000, 8000, or 8080, for example. It can be easy to make a mistake and configure your app in a way that makes it impossible for the Fly Proxy to route requests to it. And it can be difficult to debug, especially if your framework doesn't print the listening address to logs and your image doesn't have `netstat` or `ss` tools.

Author: I'll go with the small scope for now

### Example - Configure port and host in a Fastify Node app

> How do you figure out where and how your app is listening to localhost or bound to a default port? Check the code for where the web service that listens is stated up - sometimes it'll be just `serve()` or` listen()` and what's missing is parameters for the address and/or port.

Reviewer: ```suggestion
How do you figure out which address and port your app is listening on? Check your code for where the web service is started up - sometimes it'll be just `serve()` or` listen()` and what's missing is parameters for the address and/or port.
```

const fastify = Fastify({
  logger: true
})
// CommonJs
const fastify = require('fastify')({
  logger: true
})

// Declare a route
fastify.get('/', function (request, reply) {
  reply.send({ hello: 'world' })
})

// Run the server!
fastify.listen({ port: 3000 }, function (err, address) {
  if (err) {
    fastify.log.error(err)
    process.exit(1)
  }
  // Server is now listening on ${address}
})
```

This example will work locally, but when you run `fly deploy` you’ll get the “app is listening on the incorrect address” warning.

You can modify the example to listen on host `0.0.0.0` and to print a log with the listening address:

```jsx
...

fastify.listen({ port: 3000, host: '0.0.0.0' }, function (err, address) {
  if (err) {
    fastify.log.error(err)
    process.exit(1)
  }
  fastify.log.info(`server listening on ${address}`)
})
```

Then make sure that the `internal_port` value in `fly.toml` is set to `3000`. 

## Health checks failing

We don't do health checks when you launch your app for the first time, even if you use a pre-configured `fly.toml` file with health checks. But on subsequent deploys, health checks can fail for a number of reasons.

> A good first step can be to look at a failed Machine and see what you can figure out. To see the specific Machine status, run `fly status --all` to get a list of Machines in your app. Find one with state `failed` or `stopped`, then run `fly machine status <machine id>` . This will give you a lot more information. Make sure you check the exit code, if it’s 0 it means health check failures, if it’s not zero it’s some issue crashing the process.

Reviewer: ```suggestion
A good first step can be to look at a failed Machine and see what you can figure out. To see the specific Machine status, run `fly status --all` to get a list of Machines in your app. Find one with state `failed` or `stopped`, then run `fly machine status <machine id>` . This will give you a lot more information. Make sure you check the exit code: if it’s non-zero, your process crashed.
```

If a `stopped` machine has exit code 0, that could be a deliberate `fly m stop`, or an intentional process exit. Right? Also, do we actually get `failed` in `fly status`? Is that something a health check does? Wait, do health checks stop Machines these days? I haven't been adding health checks to my apps and since they don't get them by default, I'm not familiar with the behaviour anymore.

Author: Yes, health checks are a whole other topic. For now, a quick fix is to remove the bit about "stopped or failed" state. They can still look at Machine status. Will revisit this.

### Out-of-memory (OOM) or high CPU usage

If your Machine's resources are reaching their limits, then this could slow everything down, including accepting connections and responding to HTTP requests. Slow responses can trigger health check failures.

You might see out-of-memory errors in logs. Some apps (like Node.js apps that use Prisma) can be RAM-intensive. So your app may be killed for out-of-memory (OOM) reasons. The solution is to just [add more RAM](https://fly.io/docs/apps/scale-machine/#add-ram). 

If you see high CPU usage in metrics you might need to select a new [preset CPU/RAM combination](/docs/apps/scale-machine/#select-a-preset-cpu-ram-combination), or even update only the [CPU on an individual Machine](/docs/apps/scale-machine/#machines-not-belonging-to-fly-launch).

### Grace period

> Grace period is the time to wait after a Machine starts up before checking its health.

Reviewer: I don't think this is really what grace period is on Machines unless it's been improved since it was first implemented

Author: @catflydotio this came from the [fly.toml docs](https://fly.io/docs/reference/configuration/) but I can look into it

Author: So I think this is essentially right. it's the time to wait before letting flyctl make any decisions about healthy vs unhealthy Machines
